// openai/resources/beta/threads/messages.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Messages = void 0;
const resource_1 = require("../../../resource.js");
const pagination_1 = require("../../../pagination.js");
class Messages extends resource_1.APIResource {
    /**
     * Create a message.
     */
    create(threadID, body, options) {
        return this._client.post(`/threads/${threadID}/messages`, {
            body,
            ...options,
            headers: { 'OpenAI-Beta': 'assistants=v2', ...options?.headers },
        });
    }
    /**
     * Retrieve a message.
     */
    retrieve(messageID, params, options) {
        const { thread_id } = params;
        return this._client.get(`/threads/${thread_id}/messages/${messageID}`, {
            ...options,
            headers: { 'OpenAI-Beta': 'assistants=v2', ...options?.headers },
        });
    }
    /**
     * Modifies a message.
     */
    update(messageID, params, options) {
        const { thread_id, ...body } = params;
        return this._client.post(`/threads/${thread_id}/messages/${messageID}`, {
            body,
            ...options,
            headers: { 'OpenAI-Beta': 'assistants=v2', ...options?.headers },
        });
    }
    /**
     * Returns a list of messages for a given thread.
     */
    list(threadID, query = {}, options) {
        return this._client.getAPIList(`/threads/${threadID}/messages`, (pagination_1.CursorPage), {
            query,
            ...options,
            headers: { 'OpenAI-Beta': 'assistants=v2', ...options?.headers },
        });
    }
    /**
     * Deletes a message.
     */
    delete(messageID, params, options) {
        const { thread_id } = params;
        return this._client.delete(`/threads/${thread_id}/messages/${messageID}`, {
            ...options,
            headers: { 'OpenAI-Beta': 'assistants=v2', ...options?.headers },
        });
    }
}
exports.Messages = Messages;
